### PR TITLE
Add complex number support to `astype`

### PR DESCRIPTION
**Pull request by kgryte**
This PR:

-   adds complex number support to `astype`. As `astype` generally supports copying an array to _any_ specified data type irrespective of type promotion rules, no exception to general behavior is made for complex number arrays.

    Accordingly, following C, NumPy, and others, when casting a complex floating-point array to a real-valued data type, the imaginary component must be discarded and only the real component cast to the desired data type.

## Notes

-   The alternative to discarding the imaginary component is to require users call `real(x)` **before** calling `astype`. However, as `astype` is more generally used to force a type conversion, prohibiting casting complex-valued arrays to real-valued arrays did not seem ideal, nor consistent with, say, casting a floating-point array to an integral array (i.e., we allow casting a floating-point array to an integral array and don't require `astype(floor(x))`). Hence, this PR chooses to follow existing precedent and require that the imaginary component be discarded when performing the conversion.

```python
In [1]: z = np.asarray(1.5 + 2.6j)

In [2]: z.astype(dtype="float64")
<ipython-input-9-bea51b1612d9>:1: ComplexWarning: Casting complex values to real discards the imaginary part
  z.astype(dtype="float64")
Out[2]: array(1.5)

In [3]: z.astype(dtype="bool")
Out[3]: array(True)

In [4]: np.asarray(True).astype(dtype="complex128")
Out[4]: array(1.+0.j)

In [5]: np.asarray(False).astype(dtype="complex128")
Out[5]: array(0.+0.j)

In [6]: z.astype(dtype="int32")
<ipython-input-13-5f154e24a33c>:1: ComplexWarning: Casting complex values to real discards the imaginary part
  z.astype(dtype="int32")
Out[6]: array(1, dtype=int32)

In [7]: np.asarray(0+0j).astype(dtype="bool")
Out[7]: array(False)
```

## Updates

-   Based on discussions in the consortium workgroup and the 22 September 2022 workgroup meeting, consensus emerged to disallow casting from a complex floating-point dtype to a real-valued dtype. Instead, array API consumers should be explicit in specifying which component should be cast to a specified real-valued dtype.

### CODE REVIEW

**Review by leofang — COMMENTED** [review]
As long as we don't raise the annoying warning when casting complex to real, I am fine with merging this PR.

**Comment by kgryte** [comment]
@leofang Agreed. I don't think there is any appetite for requiring implementations to display a warning (ala NumPy).

**Comment by jakirkham** [comment]
Was about to say this should raise a warning 😂

That said, I kind of wonder if this is a good idea as opposed to simply forcing users to do `.real.astype(...)` instead, which seems clearer. Having `.astype(...)` shortcut this case seems prone subtle bugs in user code

What about saying `.astype(...)` behavior is undefined for types that are not complex. IOW this could still be used for upcasting from `complex64` to `complex128`

**Comment by leofang** [comment]
Quoting myself from https://github.com/data-apis/array-api/issues/463#issuecomment-1178393667:
> Based on https://github.com/data-apis/array-api/pull/427 (real) and https://github.com/data-apis/array-api/pull/446 (conj) I think we are leaning toward a more restrictive API design as opposed to permissive. I seem to remember the permissive choice was favored during one of the meetings, but I can't recall it was for which library (and I also forgot which side I picked, I'd say restrictive but I am not certain anymore after a long week...😅) That's probably why I let go in https://github.com/data-apis/array-api/pull/445 (astype). But, for the sake of consistency it'd be nicer if astype does not allow complex -> real?

**Review by leofang — CHANGES_REQUESTED** [review]

File: spec/API_specification/array_api/data_type_functions.py [hunk 1]
```diff
@@ -8,10 +8,20 @@ def astype(x: array, dtype: dtype, /, *, copy: bool = True) -> array:
        Casting floating-point ``NaN`` and ``infinity`` values to integral data types is not specified and is implementation-dependent.
 
     .. note::
-       When casting a boolean input array to a real-valued data type, a value of ``True`` must cast to a numeric value equal to ``1``, and a value of ``False`` must cast to a numeric value equal to ``0``.
+       When casting a complex floating-point array to a real-valued data type, imaginary components must be discarded (i.e., for a complex floating-point array ``x``, ``astype(x)`` must equal ``astype(real(x))``).
```
[the remark this answers is not in the document]
Comment (kgryte) [reply spec/API_specification/array_api/data_type_functions.py]: This has now been changed such that casting complex valued dtypes to real-valued dtypes is disallowed.

**Comment by kgryte** [comment]
This PR has now been updated to reflect the current workgroup consensus to disallow casting from a complex floating-point dtype to a real-valued dtype.

**Comment by kgryte** [comment]
As discussed in the 3 November 2022 meeting, no objections were raised. Will merge this PR and any further changes can be addressed in follow-up PRs.

**Review by rgommers — APPROVED** [review]
LGTM, seems ready to go in. @leofang I'll give it a few days in case you want to have another look

File: spec/API_specification/array_api/data_type_functions.py [hunk 2]
```diff
@@ -8,18 +8,28 @@ def astype(x: array, dtype: dtype, /, *, copy: bool = True) -> array:
        Casting floating-point ``NaN`` and ``infinity`` values to integral data types is not specified and is implementation-dependent.
 
     .. note::
-       When casting a boolean input array to a real-valued data type, a value of ``True`` must cast to a numeric value equal to ``1``, and a value of ``False`` must cast to a numeric value equal to ``0``.
+       Casting a complex floating-point array to a real-valued data type should not be permitted.
 
+       Historically, when casting a complex floating-point array to a real-valued data type, libraries such as NumPy have discarded imaginary components such that, for a complex floating-point array ``x``, ``astype(x)`` equals ``astype(real(x))``). This behavior is considered problematic as the choice to discard the imaginary component is arbitrary and introduces more than one way to achieve the same outcome (i.e., for a complex floating-point array ``x``, ``astype(x)`` and ``astype(real(x))`` versus only ``astype(imag(x))``). Instead, in order to avoid ambiguity and to promote clarity, this specification requires that array API consumers explicitly express which component should be cast to a specified real-valued data type.
```
Comment (leofang) [inline spec/API_specification/array_api/data_type_functions.py:13]: 👍 Zen of Python:
> There should be one-- and preferably only one --obvious way to do it.

**Review by leofang — APPROVED** [review]
Sorry I dropped the ball, LGTM!

**Comment by kgryte** [comment]
Thanks, @leofang! In it goes!

**Pull request merged**